### PR TITLE
feat(new linter): added a new linter

### LINKED ISSUE #809: Linter for space needed after right parenthesis of function symbol formals
Just notice that there seems to be no linter that complains about the following code

```r
function(x)x + 1
```

where a space is needed between `function(x)` and `x + 1`.

### PR DESCRIPTION
**Pull request by kpagacz**
Hi all,

I have added a new linter that checks for a space after the closing parenthesis of the list of formal function arguments in cases where no braces are used and the function definition spans only a single line. 

Examples:
```r
function()print("test") # BAD

function() print("test") # OK

function()
  print("test")               # OK
```

* Added a new linter that checks for a space between the right parenthesis and the function body in function definitions that span one line and don't use braces

Closes #809

### CODE REVIEW

**Comment by kpagacz** [comment]
Locally, some of the checks don't pass for me. I have branched off master, is this intended?
The checks that don't pass: 
image

image

File: R/paren_body_linter.R [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  expression <- xml2::as_list(expression)
+  Lint(
+    filename = source_file$filename,
+    message = "There should be a space between the right parenthesis and the function body.",
+    column_number = expression@col1,
+    line_number = expression@line1,
+    line = source_file$lines[[as.character(expression@line1)]],
+    ranges = list(as.integer(c(expression@col1, expression@col2)))
+  )
+}
```
Comment (kpagacz) [inline R/paren_body_linter.R:33]:
```suggestion
}

```

**Comment by AshesITR** [comment]
Hi @kpagacz,

thanks for taking the time to write a PR!

> Locally, some of the checks don't pass for me. I have branched off master, is this intended?
> The checks that don't pass:
> image
> 
> image

I had similar issues on my Windows machine.
One or more of the following may help:

1. Make sure your lintr Project directory is on your system (C:) drive and contains no spaces
2. Make sure to start the checks from a fresh R session that has no packages attached

File: R/paren_body_linter.R [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      "@col1 = preceding-sibling::OP-RIGHT-PAREN/@col1 + 1",
+      "]"
+    )
+    matched_expressions <- xml2::xml_find_all(source_file$xml_parsed_content, xpath)
+    
+    lapply(matched_expressions, get_lint_from_expression, source_file = source_file)
+  })
+}
+
+get_lint_from_expression <- function(expression, source_file) {
```
Comment (AshesITR) [inline R/paren_body_linter.R:23]: This looks very much like an alternative implementation of `xml_nodes_to_lint` in `R/utils.R`:
https://github.com/jimhester/lintr/blob/81e3b466eec949ef340e43add3207b2cca9ff86b/R/utils.R#L202-L226

Can you check if you can use that function instead?

File: R/paren_body_linter.R [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#' @describeIn linters check that there is a space between the
+#' right parenthesis and the function body when no braces are used
+#' to define a function.
+#'
+#' @export
+paren_body_linter <- function() {
+  Linter(function(source_file) {
+    if (is.null(source_file$xml_parsed_content)) return(NULL)
+
```
Comment (AshesITR) [inline R/paren_body_linter.R:9]: Can you configure your IDE to remove trailing whitespace?
There are a lot of lints due to superfluous whitespace.

File: tests/testthat/test-paren_body_linter.R [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  # paren_body_linter returns the correct line number
+  expect_lint("function()test", list(line_number = 1L), linter)
+  expect_lint(
+    "print('hello')\nx <- function(x)NULL\nprint('hello')",
+    list(line_number = 2L),
+    linter
+  )
+  
+  # paren_body_linter returns the correct type
+  expect_lint("function()test", list(type = "style"), linter)
```
Comment (AshesITR) [inline tests/testthat/test-paren_body_linter.R:27]: You can combine tests using the same sample source into one by providing multiple elements for checking to the `list()` call.

File: R/paren_body_linter.R [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    
+    lapply(matched_expressions, get_lint_from_expression, source_file = source_file)
+  })
+}
+
+get_lint_from_expression <- function(expression, source_file) {
+  expression <- xml2::as_list(expression)
+  Lint(
+    filename = source_file$filename,
+    message = "There should be a space between the right parenthesis and the function body.",
```
Comment (AshesITR) [inline R/paren_body_linter.R:27]: I suggest the message should be `"There should be a space between right parenthesis and a body expression."`.

File: R/paren_body_linter.R [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+#' to define a function.
+#'
+#' @export
+paren_body_linter <- function() {
+  Linter(function(source_file) {
+    if (is.null(source_file$xml_parsed_content)) return(NULL)
+    
+    xpath <- paste(
+      "//expr[",
+      "@line1 = preceding-sibling::FUNCTION/@line1",
```
Comment (AshesITR) [inline R/paren_body_linter.R:12]: Maybe we should also lint `if (x)bla` and `while (x)bla`?

**Review by AshesITR — CHANGES_REQUESTED** [review]
Looks good so far, just a few things need changing.

File: tests/testthat/test-paren_body_linter.R (hunk 4 above)
Comment (kpagacz) [reply tests/testthat/test-paren_body_linter.R:27]: I have done this deliberately to make the tests check only one thing at a time, but I merged it per your request.

File: R/paren_body_linter.R (hunk 6 above)
Comment (renkun-ken) [reply R/paren_body_linter.R:12]: Yes, `if`, `for`, `while`, `function`, `\` lambda are the cases where `)` must be followed by a body expression. Am I missing any?

Comment (kpagacz) [reply R/paren_body_linter.R:12]: I assume for as well.

Comment (kpagacz) [reply R/paren_body_linter.R:12]: Done

File: R/paren_body_linter.R (hunk 5 above)
Comment (kpagacz) [reply R/paren_body_linter.R]: Changed the message

File: R/paren_body_linter.R (hunk 2 above)
Comment (kpagacz) [reply R/paren_body_linter.R]: Refactored to xml_nodes_to_lint. It flew under my radar. Not all xml-based linters use this (e.g. paren_brace_linter).

File: R/paren_body_linter.R (hunk 3 above)
Comment (kpagacz) [reply R/paren_body_linter.R]: I ran lintr on the files, they are clear for me now.

File: R/paren_body_linter.R (hunk 6 above)
Comment (kpagacz) [reply R/paren_body_linter.R:12]: Oh, `\` is the new shorthand for `function()`. Added as well.

I can't think about anything else.

**Review by AshesITR — CHANGES_REQUESTED** [review]
Very nice, thank you.
I think this is ready to merge. Would you add a line to NEWS.md and update the roxygen documentation to reflect the extensions?

**Comment by kpagacz** [comment]
Done. Thanks for taking the time to review!

**Review by AshesITR — APPROVED** [review]
LGTM

**Comment by AshesITR** [comment]
@jimhester @MichaelChirico  @renkun-ken  @russHyde  do we want this in the default linters?
[It's mentioned in the tidyverse style guide](https://style.tidyverse.org/syntax.html#parentheses).

**Comment by MichaelChirico** [comment]
yes agree. IINM styler::style_file also fixes this

File: tests/testthat/test-paren_body_linter.R [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  linter <- paren_body_linter()
+  msg <- "There should be a space between right parenthesis and a body expression."
+
+  # No space after the closing parenthesis prompts a lint
+  expect_lint("function()test", msg, linter)
+  expect_lint("print('hello')\nx <- function(x)NULL\nprint('hello')", msg, linter)
+  expect_lint("if (TRUE)test", msg, linter)
+  expect_lint("while (TRUE)test", msg, linter)
+  expect_lint("for (i in seq_along(1))test", msg, linter)
+  expect_lint("\\()test", msg, linter)
```
Comment (AshesITR) [inline tests/testthat/test-paren_body_linter.R:11]: The lambda syntax should only be checked in R >= 4.0.
Curiously, the test also fails for R 4.0.

Can you investigate?

**Review by AshesITR — CHANGES_REQUESTED** [review]

File: tests/testthat/test-paren_body_linter.R (hunk 7 above)
Comment (MichaelChirico) [reply tests/testthat/test-paren_body_linter.R:11]: for failure on 4.0 itself, wasn't there a bug with utils::getParseData() on that release?

**Comment by russHyde** [comment]
I don't think unembraced blocks of the form
```
f <- function(x)
    print(x)
```
are allowed in the tidyverse style (see section "inline statements" in https://style.tidyverse.org/syntax.html#control-flow); though they are allowed if on a single line.
So this linter doesn't seem to fit with the tidyverse style, and IMO shouldn't be in the defaults

**Comment by MichaelChirico** [comment]
Russ, AIUI that example is just to show another usage that wouldn't be flagged by this linter, rather than showing code that is "OK".

That is, the code maybe should be flagged, but it would be by a different linter, unless you are proposing to extend the scope of this linter.

File: tests/testthat/test-paren_body_linter.R (hunk 7 above)
Comment (kpagacz) [reply tests/testthat/test-paren_body_linter.R:11]: I am going to look into running that particular test only for R>=4.0 and will check if the parsed tree is somehow wrong for version 4.0

**Comment by jimhester** [comment]
I think this could be added to the default linters.

**Comment by russHyde** [comment]
OK cool.

File: tests/testthat/test-paren_body_linter.R (hunk 7 above)
Comment (kpagacz) [reply tests/testthat/test-paren_body_linter.R:11]: Regarding the failure on the R version 4.0.5 (I assume that's the one running there). The new lambda syntax was introduced in the version 4.1.0 as per this NEWS document: https://cran.r-project.org/doc/manuals/r-release/NEWS.pdf (3rd point on the 3rd page).

Comment (AshesITR) [reply tests/testthat/test-paren_body_linter.R:11]: My bad re the version number. The condition needs to be on R version >= 4.1.0, not 4.0.0
Do you need help with adding the linter to the defaults?

Comment (kpagacz) [reply tests/testthat/test-paren_body_linter.R]: I have added the necessary skip condition.

File: tests/testthat/test-paren_body_linter.R [hunk 8]
```diff
@@ -43,4 +42,11 @@ testthat::test_that("paren_body_linter returns correct lints", {
 
   # multiple lints on the same line
   expect_lint("function()if(TRUE)while(TRUE)test", list(msg, msg, msg), linter)
+
+  # No space after the closing parenthesis of an anonymous function prompts a lint
+  testthat::skip_if(
+    compareVersion(paste(R.version$major, R.version$minor, sep = "."), "4.1.0") < 0,
```
Comment (AshesITR) [inline tests/testthat/test-paren_body_linter.R:48]: You can use `getRversion() >= "4.1"` since `package_version` implements a comparison.

**Review by AshesITR — CHANGES_REQUESTED** [review]

File: tests/testthat/test-paren_body_linter.R (hunk 8 above)
Comment (kpagacz) [reply tests/testthat/test-paren_body_linter.R:48]:
```suggestion
    getRversion() < "4.1",
```

**Review by AshesITR — APPROVED** [review]

**Pull request merged**